src/libcore/iter/mod.rs

```diff
@@ -1643,20 +1708,98 @@ impl<I: DoubleEndedIterator, U, F> DoubleEndedIterator for FlatMap<I, U, F> wher
 ///
 /// [`fuse()`]: trait.Iterator.html#method.fuse
 /// [`Iterator`]: trait.Iterator.html
-#[derive(Clone, Debug)]
 #[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
 #[stable(feature = "rust1", since = "1.0.0")]
-pub struct Fuse<I> {
-    iter: I,
-    done: bool
+pub struct Fuse<I: Iterator> {
+    iter: <I as SpecFuse>::T,
+}
+
+impl<I: Iterator> Fuse<I> {
+    fn new(iter: I) -> Self {
+        Fuse { iter: <I as SpecFuse>::T::new(iter) }
+    }
 }
 
+#[unstable(feature = "fused", issue = "35602")]
+impl<I> FusedIterator for Fuse<I> where I: Iterator {}
+
 #[stable(feature = "rust1", since = "1.0.0")]
-impl<I> Iterator for Fuse<I> where I: Iterator {
-    type Item = <I as Iterator>::Item;
+impl<I: Iterator + Clone> Clone for Fuse<I> {
+    fn clone(&self) -> Self {
+        Fuse { iter: self.iter.spec_clone() }
+    }
+}
 
-    #[inline]
-    fn next(&mut self) -> Option<<I as Iterator>::Item> {
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<I: Iterator + fmt::Debug> fmt::Debug for Fuse<I> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.iter.spec_debug(f)
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<I: Iterator> Iterator for Fuse<I> {
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<I::Item> {
+        self.iter.next()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<I: DoubleEndedIterator> DoubleEndedIterator for Fuse<I> {
+    fn next_back(&mut self) -> Option<I::Item> {
+        self.iter.spec_next_back()
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<I: ExactSizeIterator> ExactSizeIterator for Fuse<I> {
+    fn len(&self) -> usize {
+        self.iter.spec_len()
+    }
+}
+
+trait SpecFuse: Sized + Iterator {
+    type T: SpecFuseImpl<Iter = Self> + Iterator<Item=Self::Item>;
+}
+
+unsafe trait SpecFuseImpl: Sized {
+    type Iter: Iterator;
+
+    fn new(iter: Self::Iter) -> Self;
+
+    fn spec_next_back(&mut self) -> Option<<Self::Iter as Iterator>::Item>
+        where Self::Iter: DoubleEndedIterator;
+
+    fn spec_clone(&self) -> Self where Self::Iter: Clone;
+
+    fn spec_debug(&self, f: &mut fmt::Formatter) -> fmt::Result where Self::Iter: fmt::Debug;
+
+    fn spec_len(&self) -> usize where Self::Iter: ExactSizeIterator;
+
+    unsafe fn spec_get_unchecked(&mut self, i: usize) -> <Self::Iter as Iterator>::Item
+        where Self::Iter: TrustedRandomAccess;
+}
+
+impl<I: Iterator> SpecFuse for I {
+    default type T = RealFuse<I>;
+}
+
+#[derive(Clone)]
+struct RealFuse<I> {
+    done: bool,
```

Comment:
>  but it does have the side effect of changing the time at which the underlying iterator's destructor is called

That's why I switched to this version. The real problem is that two versions of fuse would behave very differently in an unexpected manner. For example, say your iterator holds a lock but you always use it under a `Fuse` wrapper. I'm worried that someone could write some code that relies on this dropping behavior and then try implementing `FusedIterator` for the underlying iterator. Suddenly, the inner iterator *wouldn't* be dropped which might cause a deadlock.